src/OpenTelemetry/Metrics/MetricPoint.cs

```diff
@@ -545,5 +496,40 @@ internal void TakeSnapshot(bool outputDelta)
         {
             throw new NotSupportedException($"{methodName} is not supported for this metric type.");
         }
+
+        private void UpdateHistogram(double number)
+        {
+            int i;
+            for (i = 0; i < this.histogramBuckets.ExplicitBounds.Length; i++)
+            {
+                // Upper bound is inclusive
+                if (number <= this.histogramBuckets.ExplicitBounds[i])
+                {
+                    break;
+                }
+            }
+
+            lock (this.histogramBuckets.LockObject)
+            {
+                unchecked
+                {
+                    this.runningValue.AsLong++;
+                    this.histogramBuckets.RunningSum += number;
+                    this.histogramBuckets.RunningBucketCounts[i]++;
+                }
+            }
+        }
+
+        private void UpdateHistogramSumCount(double number)
+        {
+            lock (this.histogramBuckets.LockObject)
```

Comment:
https://github.com/open-telemetry/opentelemetry-dotnet/pull/2951 showed good perf improvement when using own mechanism, as opposed to plain locks. If we modify the SnapShot path to also use the Interlock check for IsCriticalSectionOccupied, - would that not work??

cc : @utpilla

Reply:
I think we can use `Interlocked.Exchange` or `Volatile.Write` instead of simply assigning that to zero to avoid that risk.

There is an example on the docs for `Interlocked` that shows this this approach: https://docs.microsoft.com/en-us/dotnet/api/system.threading.interlocked?view=net-6.0#examples

Reply:
When I suggested using `Interlocked`, I didn't mean that we should switch every statement to use `Interlocked` like shown here:
```csharp
if (Interlocked.Exchange(ref this.histogramBuckets.IsCriticalSectionOccupied, 1) == 0)
{
    Interlocked.Increment(ref this.runningValue.AsLong);
    Interlocked.Add(ref this.histogramBuckets.RunningSum, number); // Not possible with double, but for discussion sake
    Interlocked.Increment(ref this.histogramBuckets.RunningBucketCounts[i]);

    Interlocked.Exchange(ref this.histogramBuckets.IsCriticalSectionOccupied, 0);
}
```

---

I was indeed referring to this:
```csharp
if (Interlocked.Exchange(ref this.histogramBuckets.IsCriticalSectionOccupied, 1) == 0)
{
    unchecked
    {
        this.runningValue.AsLong++;
        this.histogramBuckets.RunningSum += number;
        this.histogramBuckets.RunningBucketCounts[i]++;
    }

    Interlocked.Exchange(ref this.histogramBuckets.IsCriticalSectionOccupied, 0);
}
```

> I don't think that works. **Because the stuff in unchecked could be cached**. Needs some kind of a memory fence.

How does using a `lock` instead help with this? ^

Reply:
I think it should be okay even if the instructions inside the `unchecked` block are re-ordered.

> the stuff in unchecked could be cached

But does memory barrier/fence even help with caching/freshness?